server/web/webserver_test.go

```diff
@@ -0,0 +1,89 @@
+package web
+
+import (
+	"net/url"
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/mattermost/mattermost-server/v6/shared/mlog"
+)
+
+func Test_NewServer(t *testing.T) {
+	tests := []struct {
+		name       string
+		rootPath   string
+		serverRoot string
+		ssl        bool
+		port       int
+		localOnly  bool
+		logger     *mlog.Logger
+	}{
+		{
+			name:       "should return Server with given properties",
+			rootPath:   "./test/path/to/root",
+			serverRoot: "https://some-fake-server.com/fake-url",
+			ssl:        false,
+			port:       9999, // fake port number
+			localOnly:  false,
+			logger:     &mlog.Logger{},
+		},
+		{
+			name:       "should return local server with given properties",
+			rootPath:   "./test/path/to/root",
+			serverRoot: "https://some-fake-server.com/fake-url",
+			ssl:        false,
+			port:       3000, // fake port number
+			localOnly:  true,
+			logger:     &mlog.Logger{},
+		},
+		{
+			name:       "should match Server properties when ssl true",
+			rootPath:   "./test/path/to/root",
+			serverRoot: "https://some-fake-server.com/fake-url",
+			ssl:        true,
+			port:       8000, // fake port number
+			localOnly:  false,
+			logger:     &mlog.Logger{},
+		},
+		{
+			name:       "should return local server with given properties",
+			rootPath:   "./test/path/to/root",
+			serverRoot: "https://localhost:8080/fake-url",
+			ssl:        true,
+			port:       9999, // fake port number
+			localOnly:  true,
+			logger:     &mlog.Logger{},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			baseURL := ""
+			url, err := url.Parse(test.serverRoot)
```

Comment:
Looks like this part here is extracted from the `NewServer` code... so we are not testing the underneath behavior, we are testing that the behavior of that chunks of code are equivalent. I would suggest to add the `ExpectedBaseURL` to the test cases.

Also, here we are only testing the happy path. Actually if you look for a non-happy path, for example, passing something that is not a valid URL, you would find a bug there (If I'm right, it is going to panic in that situation).

My suggestion would be add that test cases, and fix the problem how you consider is the right approach (that can be changing the NewServer to return the server and the error, change the NewServer behavior to return nil on error (probably that is unexpected), or maybe change the NewServer to set the baseURL to a defult value (for example `/`) in that case, probably also unexpected.

Reply:
@jespino 
What would be a non-happy path be like .. 
/@#$  - in this case it returns   / 
///path  -  runs correct
/...path  - runs correct

Reply:
Sorry for the late reply, but I think you nailed it.